### ISSUE TITLE
Rendering Arabic fonts

### ISSUE THREAD

**Issue by Buccellati** [libass/libass#292]
Hi, I just switched to Libass recently and I came to realize that it doesn't render a whole lot of Arabic fonts.
Example:
Xy-Vs
![](https://i.imgur.com/BPufFZo.png)
Libass
![](https://i.imgur.com/YTY7fub.png)
This is just a drop in the bucket. Is there any way to solve this?
Link for this exact font: [Here](http://www.arfonts.net/download.php?url=mutawar-symbol,hesham-nagham-normal-traditional).

**Comment by SafaAlfulaij** [comment]
Not a libass issue, but HarfBuzz one, and I think that this type of fonts is obsolete (the definition of letters is not a standard -> shouldn't look as you would expect, better use a different and a better readable font)

**Comment by khaledhosny** [comment]
What @SafaAlfulaij said. This is a legacy pre-OpenType font that is only supported by MS text layout engine. 

HarfBuzz has some support for these fonts but it is enabled on Windows only and the client needs to use `ScriptGetCMap()` Windows API to map characters to glyphs since the fonts do not include an Arabic `cmap` subtable and Windows API seems to synthesize one.

**Comment by khaledhosny** [comment]
See https://bugzilla.mozilla.org/show_bug.cgi?id=1045139

**Comment by astiob** [comment]
Thanks for the information!

What happens in Firefox when viewing a page using such fonts on non-Windows? (The linked thread has no live links for me to see by myself, unfortunately.) Does it just use another font like libass is doing currently? It seems strange that it was done for Windows only and this raised no concerns.

In any case, if support for such fonts could be implemented on all platforms, that would be great. But if it proves too hard, we might be interested in more details about how to get it to work at least on Windows.

**Comment by khaledhosny** [comment]
Getting this to work on Windows would probably involve modifying `libass/ass_shaper.c:get_glyph()` to call [`ScriptGetCMap()`](https://msdn.microsoft.com/en-us/library/windows/desktop/dd319122%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396) (or some other Windows function) on Windows instead of Freetype functions, but this might not be trivial.

I opened https://github.com/harfbuzz/harfbuzz/issues/681 to see if HarfBuzz can support this in a cross-platform way.

**Comment by SafaAlfulaij** [comment]
> get it to work at least on Windows

And then one files a bug because it don't work on Linux, and just still open till the “coming” version of libASS is out, if ever.

**Comment by jfkthame** [comment]
AFAIK, such fonts are not supported on (most?) other platforms, so a page that calls for (e.g.) MS Sans Serif will fall back to some other font -- either the next font in the CSS font-family list, if fallbacks have been specified there, or based on the Firefox font preferences.

https://bugzilla.mozilla.org/show_bug.cgi?id=1045139 was a Windows-only bug, from Firefox's point of view, because that's the only platform where users typically have fonts like MS Sans Serif available.

**Comment by anoosi** [comment]
@khaledhosny 
can you explain how mod ass_shaper.c get_glyph() to call ScriptGetCMap()..
i using libass with harfbuzz .. some fonts like skr-head1 i recreated it with font creator ..
but must all users reinstall the font to work well with libass+harfbuzz
thanks..

**Comment by astiob** [comment]
HarfBuzz has merged support for shaping these fonts. Now we just need to figure out what to do (assuming we even need to do anything) with the cmaps on our side.

**Comment by behdad** [comment]
> HarfBuzz has merged support for shaping these fonts. Now we just need to figure out what to do (assuming we even need to do anything) with the cmaps on our side.

You shouldn't need to do anything.

**Comment by khaledhosny** [comment]
> > HarfBuzz has merged support for shaping these fonts. Now we just need to figure out what to do (assuming we even need to do anything) with the cmaps on our side.
> 
> You shouldn't need to do anything.

Unless they are using custom font functions.

**Comment by astiob** [comment]
That’s my best hope, of course, but I’m sceptical, sadly:

We have a font selection/fallback stage before HarfBuzz where we look up characters in fonts using FreeType’s API and our own logic to convert Unicode codepoints for non-Unicode cmaps. Judging by the screenshot in the first post of this thread, I suspect we skip right over the PUA font and it never even reaches HarfBuzz.

But a practical test will tell.

> Unless they are using custom font functions.

We do have custom `get_glyph_nominal` and `get_glyph_variation` if that’s what you mean. They apply the same logic I mention above that converts the Unicode codepoint parameter for non-Unicode cmaps (followed by a simple FreeType lookup). Is HarfBuzz by any chance able to do this on its own?

They also have a second job that’s a bit of our own problem, which is load, conditionally-transform and cache the glyph’s metrics, partly because HarfBuzz apparently requests the metrics a lot and caching them provides a significant speedup (maybe this is no longer the case?) and partly because we have some bad logic that rotates glyphs based on their Unicode codepoint, which we should really replace by requesting vertical writing from HarfBuzz.

**Comment by behdad** [comment]
> We have a font selection/fallback stage before HarfBuzz where we look up characters in fonts using FreeType’s API and our own logic to convert Unicode codepoints for non-Unicode cmaps. Judging by the screenshot in the first post of this thread, I suspect we skip right over the PUA font and it never even reaches HarfBuzz.

For non-Unicode cmaps, if they are common, we are interested to know and possibly implement them in HarfBuzz, like we did for this one.

You can also implement a shaper-driver font fallback scheme like Chrome and LibreOffice do, which would fix the font-selection issue here.

> They also have a second job that’s a bit of our own problem, which is load, conditionally-transform and cache the glyph’s metrics, partly because HarfBuzz apparently requests the metrics a lot and caching them provides a significant speedup (maybe this is no longer the case?)

You mean the glyph advance? I suggest you use hb-ot-font, which does all of these for you and is really fast. But if you were using it with FreeType and it was slow, hb-ft does have a cache these days.

> and partly because we have some bad logic that rotates glyphs based on their Unicode codepoint, which we should really replace by requesting vertical writing from HarfBuzz.

Yes, use our vertical writing mode please.

**Comment by Bilal2453** [comment]
Is this still waiting anything? has there been any tests on this?
I tested it on `mpv 0.35.1` and `VLC 3.0.20` and they both seem to fallback to another font. Although not sure if the HarfBuzz version has been updated on whatever used libass. (This is on Linux though not Windows)

I couldn't get the original `.rar` file linked in the original issue (seems like it expired) so I tried it with another font I know is legacy Windows 3.1.

**Comment by realfinder** [comment]
> Is this still waiting anything? has there been any tests on this? I tested it on `mpv 0.35.1` and `VLC 3.0.20` and they both seem to fallback to another font. Although not sure if the HarfBuzz version has been updated on whatever version libass those used. (This is on Linux though not Windows)
> 
> I couldn't get the original `.rar` file linked in the original issue (seems like it expired) so I tried it with another font I know is legacy Windows 3.1.
> 
> For instance, this is how the font gets rendered on latest harfbuzz with "مرحبا": ![image](https://github.com/libass/libass/assets/38175840/e69a29ac-f4e9-4b0d-a244-6091284a0228)

still not work even after this https://github.com/harfbuzz/harfbuzz/issues/681 fixed, anyway there are another tests cases in https://github.com/libass/libass/issues/633 , https://github.com/libass/libass/issues/379 and https://github.com/harfbuzz/harfbuzz/issues/2975

**Comment by Bilal2453** [comment]
Thanks for reminding me about this.
I revisited this and the problem is still happening, libass is falling to another font (Noto Sans Arabic in my case).

-----

Here are the details:
Font: [SimpArabicTest.ttf](https://github.com/harfbuzz/harfbuzz/blob/main/test/shape/data/in-house/fonts/SimpArabicTest.ttf) (especially made for testing this case)
mpv: v.0.40.0
FFmpeg backend: 7.1.1 (runtime 7.1.2)
libass: 0.17.3
harfbuzz: 11.5.1
OS: Fedora 43

(Note: those are dynamically loaded shared libraries, the versioning throughout this comment are matching.)

-----

Here is the testing scenario:
```
mpv --sub-font='/home/bilal/Downloads/fonts/SimpArabicTest.ttf' --v Test.mp4 --sub-file='/home/bilal/Downloads/fonts/Test.ass'
```

Translation file [Test.zip](https://github.com/user-attachments/files/27393154/Test.zip) .

<img width="1366" height="738" alt="Image" src="https://github.com/user-attachments/assets/2ded96cd-b473-4d7b-a5df-36154f14cc13" />

-----

Relevant logs (full logs [here](https://github.com/user-attachments/files/27392671/logs.txt)):
```
[sub/ass] fontselect: (SimpArabicTest, 700, 0) -> /home/bilal/.local/share/fonts/SimpArabicTest.ttf, 0, SimpArabicTest
[sub/ass] Glyph 0x627 not found, selecting one more font for (SimpArabicTest, 700, 0)
[sub/ass] fontselect: (SimpArabicTest, 700, 0) -> /usr/share/fonts/google-noto/NotoSansArabic-Bold.ttf, 0, NotoSansArabic-Bold
```

-----

hb-view (of the same version, I tested 3 different fonts):

<img width="724" height="565" alt="Image" src="https://github.com/user-attachments/assets/9e73367c-97d1-44b4-b02f-dab07fae07d4" />

-----

Should I open a new issue? Is this possibly something wrong with Harfbuzz? @behdad said that libass shouldn't need to do anything (beside using the newer version).

I also tried forcing a Windows-1256 charset thinking it might change how the glyphs are mapped, but it's the same thing.

[logs.txt](https://github.com/user-attachments/files/27392671/logs.txt)

**Comment by behdad** [comment]
So, the fonts render with hb-view but not in libass?

Is libass compiling HarfBuzz with `HB_TINY` or similar profile? That would explain it as it disables legacy charset support.

Or it might be just a font fallback issue whereas libass is determining that these fonts don't support Arabic, in which case the question is what is libass using for that determination.

**Comment by Bilal2453** [comment]
> So, the fonts render with hb-view but not in libass?

Yes, that's correct. `hb-view` is doing it right, libass is not; and frankly, everything else I tested on my system that uses the system Harfbuzz such as Inkscape also failed to render the font, and fell back to Noto Sans Arabic.

> Is libass compiling HarfBuzz with HB_TINY or similar profile? That would explain it as it disables legacy charset support.

No, libass itself doesn't seem to be doing that, although it is [disabling glib/cairo/freetype](https://github.com/libass/libass/blob/master/meson.build#L108-L114) related things which should be fine.

Since in this case mpv uses dynamically linked libraries, hb-view and libass are in fact using the same exact Harfbuzz build (verified with `ldd`, see below for the raw output), but also looking in to the system package for Harfbuzz, we notice that the [RPM spec](https://src.fedoraproject.org/rpms/harfbuzz/blob/rawhide/f/harfbuzz.spec) file is building the full version of Harfbuzz. *So it is likely not that*.

> Or it might be just a font fallback issue whereas libass is determining that these fonts don't support Arabic

That's more likely in my opinion, although I am not familiar with the code of libass so @astiob or someone else would have to verify this. I will try my best to go give the code 10 minutes and see if I can find it. The missing glyph most likely is the culprit, I am just not sure why is it unable to find said glyph (or which glyph is it looking for).


Thank you Behdad for your response.

<details>

<summary>ldd output of libass.so/hb-view</summary>

```
$ ldd /usr/bin/hb-view
	linux-vdso.so.1 (0x00007f83567a0000)
	libharfbuzz.so.0 => /lib64/libharfbuzz.so.0 (0x00007f8356632000)
	libharfbuzz-cairo.so.0 => /lib64/libharfbuzz-cairo.so.0 (0x00007f8356625000)
	libfreetype.so.6 => /lib64/libfreetype.so.6 (0x00007f835655b000)
	libcairo.so.2 => /lib64/libcairo.so.2 (0x00007f835641d000)
	libglib-2.0.so.0 => /lib64/libglib-2.0.so.0 (0x00007f83562c6000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f83561d1000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f8355fdc000)
	libgraphite2.so.3 => /lib64/libgraphite2.so.3 (0x00007f8355fbd000)
	libz.so.1 => /lib64/libz.so.1 (0x00007f8355f95000)
	libbz2.so.1 => /lib64/libbz2.so.1 (0x00007f8355f81000)
	libpng16.so.16 => /lib64/libpng16.so.16 (0x00007f8355f46000)
	libbrotlidec.so.1 => /lib64/libbrotlidec.so.1 (0x00007f8355f38000)
	libfontconfig.so.1 => /lib64/libfontconfig.so.1 (0x00007f8355ee6000)
	libX11.so.6 => /lib64/libX11.so.6 (0x00007f8355da1000)
	libXext.so.6 => /lib64/libXext.so.6 (0x00007f8355d8d000)
	libXrender.so.1 => /lib64/libXrender.so.1 (0x00007f8355d81000)
	libxcb.so.1 => /lib64/libxcb.so.1 (0x00007f8355d57000)
	libxcb-render.so.0 => /lib64/libxcb-render.so.0 (0x00007f8355d48000)
	libxcb-shm.so.0 => /lib64/libxcb-shm.so.0 (0x00007f8355d42000)
	libpixman-1.so.0 => /lib64/libpixman-1.so.0 (0x00007f8355c93000)
	libpcre2-8.so.0 => /lib64/libpcre2-8.so.0 (0x00007f8355be6000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f83567a2000)
	libbrotlicommon.so.1 => /lib64/libbrotlicommon.so.1 (0x00007f8355bc3000)
	libxml2.so.2 => /lib64/libxml2.so.2 (0x00007f8355a65000)
	libXau.so.6 => /lib64/libXau.so.6 (0x00007f8355a5e000)
	liblzma.so.5 => /lib64/liblzma.so.5 (0x00007f8355a29000)
$ ldd /lib64/libass.so.9
	linux-vdso.so.1 (0x00007faea3d45000)
	libm.so.6 => /lib64/libm.so.6 (0x00007faea3be8000)
	libfreetype.so.6 => /lib64/libfreetype.so.6 (0x00007faea3b1e000)
	libfribidi.so.0 => /lib64/libfribidi.so.0 (0x00007faea3afe000)
	libharfbuzz.so.0 => /lib64/libharfbuzz.so.0 (0x00007faea39ca000)
	libunibreak.so.6 => /lib64/libunibreak.so.6 (0x00007faea39a8000)
	libfontconfig.so.1 => /lib64/libfontconfig.so.1 (0x00007faea3958000)
	libc.so.6 => /lib64/libc.so.6 (0x00007faea3763000)
	/lib64/ld-linux-x86-64.so.2 (0x00007faea3d47000)
	libz.so.1 => /lib64/libz.so.1 (0x00007faea373b000)
	libbz2.so.1 => /lib64/libbz2.so.1 (0x00007faea3727000)
	libpng16.so.16 => /lib64/libpng16.so.16 (0x00007faea36ec000)
	libbrotlidec.so.1 => /lib64/libbrotlidec.so.1 (0x00007faea36de000)
	libglib-2.0.so.0 => /lib64/libglib-2.0.so.0 (0x00007faea3587000)
	libgraphite2.so.3 => /lib64/libgraphite2.so.3 (0x00007faea3566000)
	libxml2.so.2 => /lib64/libxml2.so.2 (0x00007faea3408000)
	libbrotlicommon.so.1 => /lib64/libbrotlicommon.so.1 (0x00007faea33e5000)
	libpcre2-8.so.0 => /lib64/libpcre2-8.so.0 (0x00007faea3338000)
	liblzma.so.5 => /lib64/liblzma.so.5 (0x00007faea3303000)
```

</details>

**Comment by behdad** [comment]
If fontconfig is involved, I think that *might* explain it.

**Comment by khaledhosny** [comment]
I think libass also has its own font functions, it is not using our `ot` or `ft` font functions, so it does get the symbol font handling support these two font functions implementations provide.

**Comment by behdad** [comment]
> I think libass also has its own font functions, it is not using our `ot` or `ft` font functions

Yeah that would explain it.

**Comment by Bilal2453** [comment]
Yes, indeed, I seem to be able to verify that. This entire [ass_font](https://github.com/libass/libass/blob/master/libass/ass_font.c) interface makes 0 calls to harfbuzz.

**Comment by moi15moi** [comment]
Yes, we need to do it in: https://github.com/libass/libass/blob/c425f6d7ec9ca7e5dfa3f8bbed29a6ddbf39a596/libass/ass_font.c#L255
We already know it.

But, it's not clear when GDI use this special case.

**Comment by Bilal2453** [comment]
>  We already know it.

I see. Sorry, but there didn't seem to be any issue tracking this actively. The only discussion was here was implying nothing was needed to be done.

> But, it's not clear when GDI use this special case.

So as I understand, if we were to call this function in this scenario it would be handling it right? We just need to figure out when to actually do the call?

**Comment by moi15moi** [comment]
> > But, it's not clear when GDI use this special case.
> 
> So as I understand, if we were to call this function in this scenario it would be handling it right? We just need to figure out when to actually do the call?

No, we need to know what trigger GDI to use the special arabic """encoding""" mentionned here: https://learn.microsoft.com/en-us/typography/legacy/legacy_arabic_fonts

**Comment by khaledhosny** [comment]
https://github.com/harfbuzz/harfbuzz/blob/5ee942feb6f7158f1f07e968c4d9b3adb5be07f9/src/hb-ot-cmap-table.hh#L2095-L2106
https://github.com/harfbuzz/harfbuzz/blob/5ee942feb6f7158f1f07e968c4d9b3adb5be07f9/src/hb-ot-os2-table.hh#L333-L343

**Comment by moi15moi** [comment]
@Bilal2453 If you want, you can test: https://github.com/libass/libass/pull/927

**Comment by astiob** [comment]
Here’s a reupload of the original report’s sample, plus an MKV file for testing convenience:
[libass#292 Arabic Font Test.zip](https://github.com/user-attachments/files/27569398/libass.292.Arabic.Font.Test.zip)

**Comment by Bilal2453** [comment]
As mentioned in other issues, you can find a bunch of them [here](https://www.aboaziz.net/misc/arabic%20fonts%20pack/) (mostly MCS/ACS/AGA ones).
I've tested the fix for multiple famous ones and it seems to work fine (see #927).

I am also working on a tool that uses the harfbuzz mappings to automatically modernize those fonts, it is fairly easy and it should help with things having a similar issue to libass (there are quite a few of them sadly despite using modern Harfbuzz, I think Pango/FreeType is the culprit. Edit: yes I verified it's pango). I will release it after a while and update this comment.

**Comment by Bilal2453** [comment]
I just noticed that for the original example, the middle word (third word) على is not being properly connected compared to the original way the font renders on GDI:
<img width="615" height="134" alt="image" src="https://github.com/user-attachments/assets/b6dfb460-1db9-4f18-8210-65757b2ca4bc" />

This is not a libass bug although, as Harfbuzz itself also does it like this

<img width="644" height="351" alt="Image" src="https://github.com/user-attachments/assets/8bcb3f0b-33e9-49fb-a38c-b56d3621332b" />

**Comment by moi15moi** [comment]
Report it to harfbuzz

Edit (14/05/2026): I reported it since you didn't: https://github.com/harfbuzz/harfbuzz/issues/5984
